src/Kiota.Builder/Refiners/GoRefiner.cs

```diff
@@ -421,6 +424,33 @@ private void FlattenGoFileNames(CodeElement currentElement)
         CrawlTree(currentElement, FlattenGoFileNames);
     }
 
+    private static void FixConstructorClashes(CodeElement currentElement, Func<string, string> nameCorrection)
+    {
+        var allClasses = new List<CodeClass>();
+        FindAllClasses(currentElement, allClasses);
+
+        foreach (var codeClass in allClasses)
+        {
+            if (codeClass.Name.StartsWith("New", StringComparison.OrdinalIgnoreCase))
+            {
+                var targetName = codeClass.Name.Substring(3);
+                if (allClasses.Exists(x => x.Name.Equals(targetName, StringComparison.OrdinalIgnoreCase)))
+                    codeClass.Name = nameCorrection(codeClass.Name);
+            }
+        }
+    }
+
+    private static List<CodeClass> FindAllClasses(CodeElement currentElement, List<CodeClass> codeClasses)
```

Comment:
why do it in two separate methods and allocating a potential big list?

Reply:
> why do it in two separate methods

to keep the tail recursive method clean.

> allocating a potential big list?

The entire tree is already in memory, so it's going to be a big list of references, which, usually, is cheap enough.

There are a few possible approaches, an alternative would be to collect only the classes starting with `New` but it will require 2 visits of the entire tree instead of 1.

Happy to hear about possible alternatives!

Reply:
Lgtm if you prefer this way, I don't know much about the complexity of the find methods used. :+1:

Reply:
Find can be O(1) if either of those is true:
- the element is found on the first level.
- false is passed (thanks for the callout), and it stops at the first level.

Otherwise it's about O(log(n)) since it using the child elements dictionary at each level (n being all elements in the tree)